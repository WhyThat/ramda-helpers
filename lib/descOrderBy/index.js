import {
  comparator,
  curry,
  gt,
  prop,
  sort,
} from 'ramda';

const descComparatorByProp = curry(propName => comparator(
      (a, b) => gt(prop(propName, a), prop(propName, b))));

const descOrderBy = curry((propName, list) => sort(descComparatorByProp(propName), list));

/**
 *  Sort desc a list of nested object by a prop
 *  @name descOrderBy
 *  @param {string} propName
 *  @param {Array} listOfNestedObject
 *  @returns {Array}
 *  @example
 *  descOrderBy('a', [{ a:1, b:2 }, { a:2, b:1 }])  //=> [{ a:2, b:1 }, { a:1, b:2 }]
 */
export default descOrderBy;
