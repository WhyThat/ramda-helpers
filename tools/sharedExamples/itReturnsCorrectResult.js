export default (result) => {
  test('returns correct result', () => {
    expect(result).toMatchSnapshot();
  });
};
