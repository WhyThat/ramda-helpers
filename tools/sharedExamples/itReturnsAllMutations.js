import itReturnsCorrectResult from './itReturnsCorrectResult';

export default (fn, mutations) => {
  mutations.forEach((mutation, index) => {
    const name = mutation.name || `mutations ${index}`;

    describe(name, () => {
      itReturnsCorrectResult(fn(mutation.props));
    });
  });
};
