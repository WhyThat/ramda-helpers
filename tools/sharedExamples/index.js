
import itReturnsAllMutations from './itReturnsAllMutations';
import itReturnsCorrectResult from './itReturnsCorrectResult';

export {
  itReturnsAllMutations,
  itReturnsCorrectResult,
};
