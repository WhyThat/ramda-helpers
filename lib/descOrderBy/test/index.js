import descOrderBy from '../';
import { itReturnsCorrectResult } from '../../../tools/sharedExamples';

describe('descOrderBy', () => {
  const mutations = [{
    name: 'With empty args',
    props: ['', []],
  }, {
    name: 'With a non ordered array of object',
    props: ['b', [{ a: 5, b: 1 }, { b: 4 }, { b: 2 }, { b: 3 }]],
  }, {
    name: 'With key not present in any object',
    props: ['c', [{ a: 5, b: 1 }, { b: 4 }, { b: 2 }, { b: 3 }]],
  }, {
    name: 'With key present only in few object',
    props: ['c', [{ a: 5, b: 1, c: 3 }, { b: 4, c: 6 }, { b: 2 }, { b: 3 }]],
  }];

  describe('Should be a function', () => {
    itReturnsCorrectResult(typeof descOrderBy);
  });

  mutations.forEach((mutation) => {
    describe(mutation.name, () => {
      itReturnsCorrectResult(descOrderBy(...mutation.props));
    });
  });
});
