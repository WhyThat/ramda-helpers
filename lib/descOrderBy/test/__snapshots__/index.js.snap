// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`descOrderBy Should be a function returns correct result 1`] = `"function"`;

exports[`descOrderBy With a non ordered array of object returns correct result 1`] = `
Array [
  Object {
    "b": 4,
  },
  Object {
    "b": 3,
  },
  Object {
    "b": 2,
  },
  Object {
    "a": 5,
    "b": 1,
  },
]
`;

exports[`descOrderBy With empty args returns correct result 1`] = `Array []`;

exports[`descOrderBy With key not present in any object returns correct result 1`] = `
Array [
  Object {
    "a": 5,
    "b": 1,
  },
  Object {
    "b": 4,
  },
  Object {
    "b": 2,
  },
  Object {
    "b": 3,
  },
]
`;

exports[`descOrderBy With key present only in few object returns correct result 1`] = `
Array [
  Object {
    "b": 4,
    "c": 6,
  },
  Object {
    "a": 5,
    "b": 1,
    "c": 3,
  },
  Object {
    "b": 2,
  },
  Object {
    "b": 3,
  },
]
`;
